File: Leetcode/53_Maximum Subarray/c++/solution.cpp
/*******************
 * Given an integer array nums, find the contiguous subarray (containing at least one number) which has the largest sum and return its sum.
*******************/
#include <cstdlib>
#include <cstdio>
#include <exception>
#include <vector>

using namespace std;

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        if(nums.size() < 1)
            return 0;
        
        int curSum = 0;
        int greatestSum = 0x80000000;
        for(int i = 0; i < nums.size(); i++) {
            if(curSum <= 0)
                curSum = nums[i];
            else
                curSum += nums[i];
            
            if(curSum > greatestSum)
                greatestSum = curSum;
        }
        
        return greatestSum;
    }
};

int main() {
    return 0;
}

File: Leetcode/53_Maximum Subarray/python/solution.py
import sys

class Solution:
    def maxSubarray(self, nums):
        if len(nums) < 1:
            return 0

        cursum = 0
        greatestSum = -sys.maxsize

        for i in range(0, len(nums)):
            if cursum <= 0:
                cursum = nums[i]
            else:
                cursum += nums[i]

            if cursum > greatestSum:
                greatestSum = cursum

        return greatestSum